### PR TITLE
Improved the synchronization by using a wait mechanism

### PR DESCRIPTION
**Pull request by judax**
Only send AR data if a camera frame is rendered. Only render a camera frame if the JS side used the AR data. Also made the WebView of the size of the view becase performance improves and because there is a disparity between the camera feed size and the webview content size. UI for the navigation bar needs to be revisited/redone.

### CODE REVIEW

File: WebARonARKit/WebARonARKit.js [hunk 1]
```diff
@@ -968,6 +972,8 @@
     }
 
     callRafCallbacks();
+    
+    window.webkit.messageHandlers.WebARonARKit.postMessage("arDataWasUsed:");
```
Comment (lincolnfrog) [inline WebARonARKit/WebARonARKit.js]: can we call this advanceFrame? I think it's clearer and less weird

File: WebARonARKit/WebARonARKit/ViewController.m [hunk 2]
```diff
@@ -290,9 +297,10 @@ - (void)viewDidLoad
     // Create the WKWebView using the configuration/script injection and add it to
     // the top of the view graph
     self->wkWebView = [[WKWebView alloc]
-        initWithFrame:CGRectMake(
-                          0, URL_TEXTFIELD_HEIGHT, self.view.frame.size.width,
-                          self.view.frame.size.height - URL_TEXTFIELD_HEIGHT)
+//        initWithFrame:CGRectMake(
```
Comment (lincolnfrog) [inline WebARonARKit/WebARonARKit/ViewController.m]: delete instead of just comment out?

File: WebARonARKit/WebARonARKit/ViewController.m [hunk 3]
```diff
@@ -441,8 +449,9 @@ - (void)deviceOrientationDidChange:(NSNotification *)notification
                     URL_TEXTFIELD_HEIGHT, URL_TEXTFIELD_HEIGHT)];
 
     [self->wkWebView setFrame:
-            CGRectMake(0, URL_TEXTFIELD_HEIGHT, self.view.frame.size.width,
-                    self.view.frame.size.height - URL_TEXTFIELD_HEIGHT)];
+//            CGRectMake(0, URL_TEXTFIELD_HEIGHT, self.view.frame.size.width,
+//                    self.view.frame.size.height - URL_TEXTFIELD_HEIGHT)];
```
Comment (lincolnfrog) [inline WebARonARKit/WebARonARKit/ViewController.m]: delete instead of just comment out?

File: WebARonARKit/WebARonARKit.js (hunk 1 above)
Comment (rezaali) [reply WebARonARKit/WebARonARKit.js]: +1

File: WebARonARKit/WebARonARKit/ViewController.m (hunk 2 above)
Comment (rezaali) [reply WebARonARKit/WebARonARKit/ViewController.m]: +1

File: WebARonARKit/WebARonARKit/ViewController.m (hunk 3 above)
Comment (rezaali) [reply WebARonARKit/WebARonARKit/ViewController.m]: +1

File: WebARonARKit/WebARonARKit/ViewController.m [hunk 4]
```diff
@@ -1016,6 +1060,10 @@ - (void)userContentController:(WKUserContentController *)userContentController
           [objCAnchorIdsToJSAnchorIds removeObjectForKey:objCAnchorId];
           [anchors removeObjectForKey:jsAnchorId];
           [self.session removeAnchor:anchor];
+        } else if ([method isEqualToString:@"arDataWasUsed"]) {
```
Comment (rezaali) [inline WebARonARKit/WebARonARKit/ViewController.m]: Just echoing that "advanceFrame" is more ideal than "arDataWasUsed"

**Comment by judax** [comment]
Added a patch responding to some of the comments. Also added support to the polyfill for both our iOS and Android approaches.

**Pull request merged**